### ISSUE TITLE
pathlib.Path: glob and rglob should accept PathLike patterns

### ISSUE THREAD

**Issue by 6d8be2ff-9207-41af-b7aa-5f3271a2e139** [python/cpython#79634]
BPO | [35453](https://bugs.python.org/issue35453)
--- | :---
Nosy | @serhiy-storchaka, @nanjekyejoannah, @isidentical, @ciupicri

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2018-12-10.10:54:42.124>
labels = ['3.8', 'type-feature', 'library']
title = 'pathlib.Path: glob and rglob should accept PathLike patterns'
updated_at = <Date 2019-11-15.23:43:31.989>
user = 'https://github.com/ciupicri'
```

bugs.python.org fields:
```python
activity = <Date 2019-11-15.23:43:31.989>
actor = 'BTaskaya'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Library (Lib)']
creation = <Date 2018-12-10.10:54:42.124>
creator = 'ciupicri'
dependencies = []
files = []
hgrepos = []
issue_num = 35453
keywords = []
message_count = 8.0
messages = ['331491', '331503', '331507', '331567', '343840', '356682', '356684', '356728']
nosy_count = 4.0
nosy_names = ['serhiy.storchaka', 'nanjekyejoannah', 'BTaskaya', 'ciupicri']
pr_nums = []
priority = 'normal'
resolution = None
stage = 'test needed'
status = 'open'
superseder = None
type = 'enhancement'
url = 'https://bugs.python.org/issue35453'
versions = ['Python 3.8']
```

</p></details>

### Linked PRs
* gh-114017
* gh-114623

**Comment by barneygale** [comment]
I'm going to disagree with Serhiy of 4 years ago (sorry!) and argue that this bug is valid. Glob patterns are path-like in a conceptual sense and a practical sense.

Conceptually they're paths with wildcards. They can have drives and roots, directories, filenames, parents, etc. All paths are valid patterns, and most paths are barely modified in order to transform them into a pattern (e.g. file `dir/a.txt` -> pattern `dir/*.txt`).

Practically __pathlib parses these patterns as paths__, e.g. it normalizes alternate separators, splits into drive/root/tail, collapses empty segments and '.' segments. Pretty much the _only_ thing it doesn't do is call `os.fspath()`.

> '*.txt' is aт ordinary file name on Linux and is invalid file name on Windows. If we wanted to accept Path in any place that accepts a string, we would make Path a subclass of str.

pathlib won't complain if you run `pathlib.Path('*.txt')` on any platform!

**Comment by barneygale** [comment]
In fact, the original bug report is no longer reproducible. It looks like it was "fixed" in 187949ebf2ae36fcf0817a06f4a7637d0a8b7fc5, which made `_parse_parts()` use `os.path.join()`.

**Comment by barneygale** [comment]
This is blocked on #65238.

**Comment by barneygale** [comment]
Resolved in 3.13 / [6313cdd](https://github.com/python/cpython/commit/6313cdde58f34648a430d2830357c9d2a5b67b87) / #114017
